core/src/main/java/io/opencensus/tags/TagValue.java

```diff
@@ -0,0 +1,182 @@
+import io.opencensus.tags.TagKey.TagKeyLong;
+import io.opencensus.tags.TagKey.TagKeyString;
+import javax.annotation.concurrent.Immutable;
+
+/** A validated tag value. */
+@Immutable
+public abstract class TagValue {
+
+  TagValue() {}
+
+  /**
+   * Applies a function to a tag value. The function that is called depends on the type of the
+   * value. This is similar to the visitor pattern. {@code match} also takes a function to handle
+   * the default case, for backwards compatibility when tag types are added.
+   *
+   * @param stringFunction the function to call when the tag value has type {@code String}.
+   * @param longFunction the function to call when the tag value has type {@code long}.
+   * @param booleanFunction the function to call when the tag value has type {@code boolean}.
+   * @param defaultFunction the function to call when the tag value has a type other than {@code
+   *     String}, {@code long}, or {@code boolean}.
+   * @param <T> The result type of the function.
+   * @return The result of calling the function that matches the tag value's type.
+   */
+  public abstract <T> T match(
+      Function<? super TagValueString, T> stringFunction,
+      Function<? super TagValueLong, T> longFunction,
+      Function<? super TagValueBoolean, T> booleanFunction,
+      Function<? super TagValue, T> defaultFunction);
+
+  /**
+   * A validated tag value associated with a {@link TagKeyString}.
+   *
+   * <p>Validation ensures that the {@code String} has a maximum length of {@link #MAX_LENGTH} and
+   * contains only printable ASCII characters.
+   */
+  @Immutable
+  @AutoValue
+  public abstract static class TagValueString extends TagValue {
+    /** The maximum length for a {@code String} tag value. The value is {@value #MAX_LENGTH}. */
+    public static final int MAX_LENGTH = 256;
+
+    TagValueString() {}
+
+    /**
+     * Constructs a {@code TagValueString} from the given string. The string must meet the following
+     * requirements:
+     *
+     * <ol>
+     *   <li>It cannot be longer than {@link #MAX_LENGTH}.
+     *   <li>It can only contain printable ASCII characters.
+     * </ol>
+     *
+     * @param value the tag value.
+     * @throws IllegalArgumentException if the {@code String} is not valid.
+     */
+    public static TagValueString create(String value) {
+      Preconditions.checkArgument(isValid(value));
+      return new AutoValue_TagValue_TagValueString(value);
+    }
+
+    @Override
+    public final <T> T match(
+        Function<? super TagValueString, T> stringFunction,
+        Function<? super TagValueLong, T> longFunction,
+        Function<? super TagValueBoolean, T> booleanFunction,
+        Function<? super TagValue, T> defaultFunction) {
+      return stringFunction.apply(this);
+    }
+
+    /**
+     * Returns the tag value as a {@code String}.
+     *
+     * @return the tag value as a {@code String}.
+     */
+    public abstract String asString();
```

Comment:
Why 'as' rather then 'get' here and below?

Reply:
The only reason I used asString/asLong/asBoolean is that Tag already has a `getValue()` method, so a user would have to call two different `getValue`s to get the unwrapped value from a Tag, i.e., `tag.getValue().getValue()`.